### PR TITLE
TS-5088: Fix missing initial buffer from server, when combined with 101 response

### PR DESCRIPTION
**Pull request by mervynmcmurray**
Author: Mervyn McMurray <mervyn.mcmurray@cisco.com>
Tested: Linux 4.10.14

### CODE REVIEW

File: proxy/http/HttpSM.h [hunk 1]
```diff
@@ -473,7 +473,7 @@ class HttpSM : public Continuation
   void perform_cache_write_action();
   void perform_transform_cache_write_action();
   void perform_nca_cache_action();
-  void setup_blind_tunnel(bool send_response_hdr);
+  void setup_blind_tunnel(bool send_response_hdr, IOBufferReader *initial=0);
```
Comment (scw00) [inline proxy/http/HttpSM.h]: Could we use `nullptr` instead !

**Comment by scw00** [comment]
[approve ci]

File: proxy/http/HttpSM.cc [hunk 2]
```diff
@@ -1593,8 +1593,12 @@ HttpSM::handle_api_return()
     if (unlikely(t_state.did_upgrade_succeed)) {
       // We've successfully handled the upgrade, let's now setup
       // a blind tunnel.
+      IOBufferReader *initial_data = 0;
```
Comment (scw00) [inline proxy/http/HttpSM.cc]: I'd rather use C++ style, The `nullptr` may be better !

File: proxy/http/HttpSM.h (hunk 1 above)
Comment (mervynmcmurray) [inline proxy/http/HttpSM.h]: I Have made the changes - need to run through tests to ensure no regression (even though it is extremely unlikely)

Comment (scw00) [inline proxy/http/HttpSM.h]: :->

**Comment by scw00** [comment]
[approve ci]

**Comment by bryancall** [comment]
@scw00 Please shepherd this PR.

File: proxy/http/HttpSM.cc [hunk 3]
```diff
@@ -6626,6 +6630,9 @@ HttpSM::setup_blind_tunnel(bool send_response_hdr)
   milestones[TS_MILESTONE_SERVER_BEGIN_WRITE] = Thread::get_hrtime();
   if (send_response_hdr) {
     client_response_hdr_bytes = write_response_header_into_buffer(&t_state.hdr_info.client_response, to_ua_buf);
+    if (initial && initial->read_avail()) {
```
Comment (scw00) [inline proxy/http/HttpSM.cc]: For `CONNECT` method, I think we also need to transfer the data(in server_reader_buffer) after sending response header. So, could we move it out of block and remove the changes in handle_api_return ?

Finally do not forget to consume data just like HttpSM::server_transfer_init

Comment (mervynmcmurray) [inline proxy/http/HttpSM.cc]: I have added the consume data as suggested. I will need more context around the CONNECT method code change. 
When developing the fix I had a very targeted bug which was easy to reproduce, hence I only fixed the initial buffer scenario. My automated test will not cover extension to the CONNECT method, but I will be able to determine if the suggested change breaks things.

Comment (scw00) [inline proxy/http/HttpSM.cc]: That's correct, just fix the the websocket on this pr.

**Comment by scw00** [comment]
Squash please :-)

**Comment by scw00** [comment]
[approve ci]

**Comment by scw00** [comment]
[approve ci]

**Comment by scw00** [comment]
[approve ci autest]

**Comment by scw00** [comment]
[approve ci autest]

**Comment by scw00** [comment]
[approve ci autest]

**Comment by scw00** [comment]
[approve ci]

**Comment by scw00** [comment]
[approve linux]

**Comment by scw00** [comment]
[approve ci autest]

**Comment by scw00** [comment]
[approve ci linux]

**Comment by scw00** [comment]
[approve ci fedora]

**Comment by scw00** [comment]
[approve ci fedora]

**Pull request merged**